evalml/utils/nullable_type_utils.py

```diff
@@ -0,0 +1,103 @@
+import woodwork as ww
+
+
+def _downcast_nullable_X(X, handle_boolean_nullable=True, handle_integer_nullable=True):
```

Comment:
There's a lot of similar logic here between this and `_downcast_nullable_y`. I think it would make sense to combine these into a single function, and separate the X/y logic based on whether the input is a DataFrame or a Series. 

It feels a bit more flexible to have a single function rather than two separate ones, and it would be able to condense down the number of tests in `test_nullable_type_utils.py`, since there's a lot of repeated logic between X and y tests there as well.

Reply:
I don't disagree about the shared logic. I generally like utilities to not do too many things (it helps me with readability of code and how I think about tests but does, indeed, end up with more tests, and I totally see where you're coming from that these tests are very similar.)

I'm gonna play around with different ways of sharing more logic among the separate utils vs combining them and see if I can put something together that doesn't feel as repetitive. One thing may be to test the shared logic separately, so that the tests for the two utils are really only testing the differing apis for series vs dataframes.

Reply:
@eccabay I pulled out more shared logic in the utils (along with some variable renaming) and refactored some tests to not be so repetitive across X and y downcasters: https://github.com/alteryx/evalml/pull/4024/commits/601535b3826ff0c85592daf9722640639657f688. I still feel like it makes sense to not squish them into one util with an if/else block since the woodwork dataframe and series apis are different enough that the two blocks of code wouldn't change much if they were in a single util at this point. It just puts the onus on anyone trying to understand the util in the future to understand the scope of the util's abilities.

If we decide to go with one util in the end, I would keep the tests the same and just remove the `downcast_util` parameter in fixtures, so making the change to one util will be very quick, and I can definitely be convinced that that's the right path.

I chose not to refactor tests that would need branching logic at both the setup and assertions, as I feel like that is a good indicator of when two checks really deserve their own tests.